server/app/models.py

```diff
@@ -658,15 +660,50 @@ def mark_as_final(self):
             final = FinalSubmission.query(
                 FinalSubmission.assignment==assignment,
                 FinalSubmission.submitter==submitter).get()
+        return final
+
+    def mark_as_final(self):
+        """Create or update a final submission."""
+        final = self.get_final()
+
         if final:
             final.submitter = self.submitter
             final.submission = self.key
         else:
-            final = FinalSubmission(assignment=assignment, submission=self.key)
+            group = self.submitter.get().get_group(self.assignment)
+            final = FinalSubmission(
+                assignment=self.assignment, submission=self.key)
             if group:
                 final.group = group.key
         final.put()
 
+    def resubmit(self, user_key):
```

Comment:
I'm not worried about bugs in this function- I'm more concerned with other things that can happen on submission (e.g. autograding) that may or may not happen here. I think for now, this is good, but in the future, we should get the actual submission system to have a hook that does what we want it to do. It's not very high priority, so I'm fine with just leaving a TODO for the future.